### ISSUE TITLE
Allow extending native types

### ISSUE THREAD

**Issue by matjaz** [firebase/bolt#21]
```
type PositiveInteger extends number {
  validate() { return this >= 0; }
}

type UnixTimestamp extends PositiveInteger {}

type NonEmptyString extends string {
  validate() { return this.length > 0; }
}

type URL extends string {
  validate() { return this.beginsWith('http'); }
}

type Test {
  time: UnixTimestamp
  name: NonEmptyString
  url: URL
}


path / {
  validate() { return this instanceof Test; }
}
```

Curently this translates to
`".validate": "newData.child('time').isNumber() && newData.child('time').val() >= 0 && (newData.child('name').isString() && newData.child('name').child('length').val() > 0) && (newData.child('url').isString() && newData.child('url').child('beginsWith')('http'))"` which is incorrect.

**Comment by matjaz** [comment]
Current workaround is to add `.val()`

```
type PositiveInteger extends number {
  validate() { return this >= 0; }
}

type UnixTimestamp extends PositiveInteger {}

type NonEmptyString extends string {
  validate() { return this.val().length > 0; }
}

type URL extends string {
  validate() { return this.val().beginsWith('http'); }
}

type Test {
  time: UnixTimestamp
  name: NonEmptyString
  url: URL
}


path / {
  validate() { return this instanceof Test; }
}
```

**Comment by mckoss** [comment]
I propose to fix this by replacing:

this.length -> length(this)
this.beginsWith -> beginsWith(this)

in Bolt expressions.

I want to add these global functions for sure:

length(string)
includes(string, substring) -> string.contains(substring)
startsWith(string, substring) -> string.beginsWith(substring)
endsWith(string, substring) ->  string.endsWith(substring)
replace(string, oldStr, newStr) -> string.replace(oldStr, newStr)
toLowerCase(string) -> string.toLowerCase()
toUpperCase(string) -> string.toUpperCase()
test(string, regexp) -> string.matches(regexp)

parent(location) -> location.parent()

I don't think the following need to be global function in Bolt since they are inferred from the type specification or are replaced with other Bolt constructs:

isNumber(), isString(), isBoolean() - implicit in type declarations

val() - implicit when location used in expression
child() - location.child or location[string]
hasChild() - location.child != null
hasChildren() - implicit in type declarations or location.child1 != null && location.child2 != null ...
exists() - location != null

**Comment by matjaz** [comment]
IMHO I don't like globals. I would more likely see proper static analysis and add methods to (parent) types (example above).
With globals you pollute global namespace, that means when you add new global function to bolt you are risking users will have conflict.
I would try to remove global functions and prefer methods via mixins and Type static methods.

``` typescript
type User {
  static exists(id) {
    return root.users[id].exists();
  }
}
```

I think there should be as less as possible globals (from https://www.firebase.com/docs/security/quickstart.html).
One I think of:
- now
- root
- newData (= this)
- data
- $var
- auth

**Comment by mckoss** [comment]
The reason for the function notation, is that we don't collide with
user-defined properties and still allow the much easier to read
dot-notation for child properties.

I think there will always be "ambient" functions in the environment of the
security engine.  Perhaps we should namespace any global functions into
"packages" for string.* and db.* (at the expense of more typing ... which I
know you hate :-) ).

    string.beginsWith(myString, substring)

On Fri, Aug 28, 2015 at 12:35 PM, Matjaž Lipuš <notifications@github.com>
wrote:

> IMHO I don't like globals. I would more likely see proper static analysis
> and add methods to (parent) types (example above).
> With globals you pollute global namespace, that means when you add new
> global function to bolt you are risking users will have conflict.
> I would try to remove global functions and prefer methods via mixins and
> Type static methods.
>
> type User {
>   static exists(id) {
>     return root.users[id].exists();
>   }
> }
>
> I think there should be as less as possible globals (from
> https://www.firebase.com/docs/security/quickstart.html).
> One I think of:
>
>    - now
>    - root
>    - newData (= this)
>    - data
>    - $var
>    - auth
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/firebase/bolt/issues/21#issuecomment-135868176>.
>



-- 
+Mike Koss <https://plus.google.com/+MikeKoss>
(425) 246-7701 (m)

**Comment by tomlarkworthy** [comment]
global functions are not a nice solution. String.beginsWith namespaceing is too verbose IMHO.

Given user defined properties are *properties* NOT functions. I think its possible to disambiguate the use by on whether there is a function call or not succeeding the identifier*.

data.length    // data.child('length') 
data.length() // data.val().toString().length()

* I suspect the implementation is obnoxious.

**Comment by matjaz** [comment]
Actually it's more like
```
data.length // data.child('length').val()
data.length() // data.val().length
```
and yeah implementation won't be trivial.

**Comment by mckoss** [comment]
I'll have to think about this.  If might not be too hard because we have good type information - and all function expansion is done in a specific context (so even one function might expand .length differently depending on the context of the caller's type).

The internal asp does not bind to a specific implementation of a property reference - the conversion to JSON is done late-bound.

**Comment by mckoss** [comment]
Fixed by f1b28f34d7bc48310057f239b026c28d19ebc11c
